### PR TITLE
[MSEARCH-784] add bibframe authority index

### PR DESCRIPTION
**Pull request by aleksei-pronichev**
### Purpose
 add bibframe authority index for mod-linked-data module

### Approach
add a new kafka topic with new data for bibframe authority index

### Changes Checklist
- add new endpoint /search/bibframe/authority
- add kafka-listener for search.bibframe-authorities topic
- 
### Related Issues
[MSEARCH-784] (https://folio-org.atlassian.net/browse/MSEARCH-784)

### CODE REVIEW

File: src/main/java/org/folio/search/service/setter/bibframe/authority/BibframeAuthorityLccnProcessor.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import org.folio.search.service.lccn.LccnNormalizer;
+import org.folio.search.service.setter.FieldProcessor;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Component;
+
+@Component
+@RequiredArgsConstructor
+public class BibframeAuthorityLccnProcessor implements FieldProcessor<BibframeAuthority, Set<String>> {
+
+  @Qualifier("lccnNormalizerStructureB")
```
Comment (pkjacob) [inline src/main/java/org/folio/search/service/setter/bibframe/authority/BibframeAuthorityLccnProcessor.java:23]: Hi @aleksei-pronichev, We have to use the DefaultLccnNormalizer here. 
lccnNormalizerStructureB is only applicable for Bib records acquired after 2001

File: src/main/java/org/folio/search/integration/KafkaMessageListener.java [hunk 2]
```diff
@@ -212,6 +213,22 @@ public void handleBibframeEvents(List<ConsumerRecord<String, ResourceEvent>> con
...
   }
 
+  @KafkaListener(
+    id = KafkaConstants.BIBFRAME_AUTHORITY_LISTENER_ID,
+    containerFactory = "standardListenerContainerFactory",
+    groupId = "#{folioKafkaProperties.listener['bibframe-authorities'].groupId}",
+    concurrency = "#{folioKafkaProperties.listener['bibframe-authorities'].concurrency}",
+    topicPattern = "#{folioKafkaProperties.listener['bibframe-authorities'].topicPattern}")
+  public void handleBibframeAuthorityEvents(List<ConsumerRecord<String, ResourceEvent>> consumerRecords) {
+    log.info("Processing bibframe events from Kafka [number of events: {}]", consumerRecords.size());
```
Comment (pkjacob) [inline src/main/java/org/folio/search/integration/KafkaMessageListener.java:223]: Processing bibframe **authority** events ...

**Comment by pkjacob** [comment]
Hi @aleksei-pronichev, Let us add a new line in News.md file (line 22)
* Create bibframe authority index and process bibframe authority events ([MSEARCH-784](https://folio-org.atlassian.net/browse/MSEARCH-784))

File: descriptors/ModuleDescriptor-template.json [hunk 3]
```diff
@@ -111,6 +111,15 @@
             "search.bibframe.collection.get"
           ]
         },
+        {
```
Comment (pkjacob) [inline descriptors/ModuleDescriptor-template.json:114]: Let us add this API to README file
| GET    | `/search/bibframe/authorities `            | Search linked data graph authority resource descriptions        |

File: src/test/resources/samples/bibframe/authority_concept.json [hunk 4]
```diff
@@ -0,0 +1,15 @@
+{
+  "id": "123456654321",
+  "label": "concept-label",
+  "type": "CONCEPT",
+  "identifiers": [
+    {
+      "value": "2024-202454/AC/r967",
```
Comment (pkjacob) [inline src/test/resources/samples/bibframe/authority_concept.json:7]: This is not the correct format for Authority LCCNs.
For CONCEPT resources, let us use the following LCCNs
sh1234557890
sh 0123456789 (note that there is a space after "sh". This should get removed by the DefaultLccnNormalizer)

File: src/test/resources/samples/bibframe/authority_person.json [hunk 5]
```diff
@@ -0,0 +1,11 @@
+{
+  "id": "123456654322",
+  "label": "person-label",
+  "type": "PERSON",
+  "identifiers": [
+    {
+      "value": "2024-202454/PR/r006",
```
Comment (pkjacob) [inline src/test/resources/samples/bibframe/authority_person.json:7]: For PERSON, let us use the LCCN "n 1234567890" (note that there is a space after "n". This should get removed by the DefaultLccnNormalizer)

File: descriptors/ModuleDescriptor-template.json [hunk 6]
```diff
@@ -111,6 +111,15 @@
...
+          "methods": [
+            "GET"
+          ],
+          "pathPattern": "/search/bibframe/authorities",
+          "permissionsRequired": [
+            "search.bibframe.collection.get"
```
Comment (mukhiddin-yusuf) [inline descriptors/ModuleDescriptor-template.json:120]: was it intentional to reuse the same permission which was also for ```/search/bibframe``` or should here be new bibframe authority collection get permission?

**Review by PBobylev — APPROVED** [review]

File: descriptors/ModuleDescriptor-template.json (hunk 6 above)
Comment (aleksei-pronichev) [reply descriptors/ModuleDescriptor-template.json:120]: yes, we decided that we don't need separate permission for our module and this one can be reused

File: descriptors/ModuleDescriptor-template.json [hunk 7]
```diff
@@ -111,6 +111,15 @@
             "search.bibframe.collection.get"
           ]
         },
+        {
+          "methods": [
+            "GET"
+          ],
+          "pathPattern": "/search/bibframe/authorities",
```
[the remark this answers is not in the document]
Comment (pkjacob) [reply descriptors/ModuleDescriptor-template.json:118]: This change will break the UI. We will create a different ticket for this and add to next sprint.

File: src/main/java/org/folio/search/utils/SearchUtils.java [hunk 8]
```diff
@@ -39,6 +39,7 @@ public class SearchUtils {
   public static final String LOCATION_RESOURCE = "location";
   public static final String CLASSIFICATION_TYPE_RESOURCE = "classification-type";
   public static final String BIBFRAME_RESOURCE = "bibframe";
+  public static final String BIBFRAME_AUTHORITY_RESOURCE = "bibframe-authority";
```
[the remark this answers is not in the document]
Comment (pkjacob) [reply src/main/java/org/folio/search/utils/SearchUtils.java:42]: Renaming "bibframe" to "bibframe-work" will cause a change in kafka topic name & break the application. So, we will address this in next sprint as part of another story.

File: src/main/resources/application.yml [hunk 9]
```diff
@@ -139,6 +139,10 @@ folio:
         concurrency: ${KAFKA_BIBFRAME_CONCURRENCY:1}
         topic-pattern: (${folio.environment}\.)(.*\.)search\.bibframe
         group-id: ${folio.environment}-mod-search-bibframe-group
+      bibframe-authorities:
+        concurrency: ${KAFKA_BIBFRAME_CONCURRENCY:1}
+        topic-pattern: (${folio.environment}\.)(.*\.)search\.bibframe-authorities
```
[the remark this answers is not in the document]
Comment (pkjacob) [reply src/main/resources/application.yml:144]: We will need a different consumer due to the different structure of messages.

[the remark this answers is not in the document]
Comment (psmagin) [reply src/main/resources/application.yml:144]: But I can't see difference in processing:
image
Both call `indexResources` method the only difference is in setting resource name and id
Please check `ResourceEventBatchInterceptor` if it's logic is acceptable for your needs.

File: src/main/java/org/folio/search/integration/KafkaMessageListener.java
Comment (psmagin) [inline src/main/java/org/folio/search/integration/KafkaMessageListener.java:1]: Please update KafkaMessageListenerTest

File: descriptors/ModuleDescriptor-template.json (hunk 6 above)
Comment (pkjacob) [reply descriptors/ModuleDescriptor-template.json:120]: We will create a new permission `search.bibframe.authority.collection.get`
We will also rename the existing permission to `search.bibframe.work.collection.get` as part of [MSEARCH-786](https://folio-org.atlassian.net/browse/MSEARCH-786)

File: src/main/resources/model/bibframe_authority.json [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+        }
+      }
+    }
+  },
+  "searchFields": {
+    "label": {
+      "type": "search",
+      "index": "multilang",
+      "processor": "bibframeAuthorityLabelProcessor"
+    },
```
Comment (psmagin) [inline src/main/resources/model/bibframe_authority.json:32]: I think it will override the `label` field in the `fields` block as the names are the same. Maybe it makes sense to set index=multilang for label in fields block.

File: src/main/resources/model/bibframe_authority.json [hunk 11]
```diff
@@ -0,0 +1,45 @@
+{
+  "name": "bibframe-authority",
+  "eventBodyJavaClass": "org.folio.search.domain.dto.BibframeAuthority",
+  "languageSourcePaths": ["$.languages"],
```
Comment (psmagin) [inline src/main/resources/model/bibframe_authority.json:4]: You specify that your schema has `languages` field that could be used then for multilang index. But I can't see the field in bibframeAuthority schema.
If there is no such field in schema then it doesn't have sense to use multilang index.

File: descriptors/ModuleDescriptor-template.json (hunk 6 above)
Comment (mukhiddin-yusuf) [reply descriptors/ModuleDescriptor-template.json:120]: @pkjacob , is there link to some UXPROD feature (or wiki page) that describes the feature and changes being added in the scopes of MSEARCH-784 and MSEARCH-786? I am asking just to understand more about the context as I could not find more info about it and there is no relations between the stories. Looking at [MSEARCH-786](https://folio-org.atlassian.net/browse/MSEARCH-786) I am wondering maybe some better names can be suggested for the API, topic, index and permissions as from the readability point of view I did not get what ```"bibframe-work"``` could mean

Comment (pkjacob) [reply descriptors/ModuleDescriptor-template.json:120]: Hi @mukhiddin-yusuf , Here is the UXPROD feature -> https://folio-org.atlassian.net/browse/UXPROD-4672. The details given in the feature story is very high level. So, I am not sure if you will get the information you are looking for from this feature story.

There are lot of articles in the internet (mainly from Library of Congress) that explains Bibframe & its various components. Here are some documents.

https://www.loc.gov/bibframe/docs/bibframe2-model.html
https://en.wikipedia.org/wiki/BIBFRAME
https://www.loc.gov/bibframe/docs/bibframe2-model.html
https://www.librarianshipstudies.com/2017/12/bibframe.html


In summary,
[MSEARCH-781](https://folio-org.atlassian.net/browse/MSEARCH-781) creates an index for Bibframe Work resources 
[MSEARCH-784](https://folio-org.atlassian.net/browse/MSEARCH-784) creates an index for Bibrame Authority resources.

 [MSEARCH-786](https://folio-org.atlassian.net/browse/MSEARCH-786) is a clean up story to fix some inconsistencies in these two indexes.

File: src/main/resources/application.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (aleksei-pronichev) [reply src/main/resources/application.yml:144]: get idea, thank you

**Review by pkjacob — APPROVED** [review]

**Comment by pkjacob** [comment]
Hi @aleksei-pronichev, There are some merge conflicts with master branch. Let us address it.

File: src/main/java/org/folio/search/service/setter/bibframe/authority/BibframeAuthorityTypeProcessor.java [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+@Component
+public class BibframeAuthorityTypeProcessor implements FieldProcessor<BibframeAuthority, String> {
+
+  @Override
+  public String getFieldValue(BibframeAuthority bibframe) {
+    var type = bibframe.getType();
+    if (type == null) {
+      return null;
+    }
+    return type.toString();
```
Comment (pkjacob) [inline src/main/java/org/folio/search/service/setter/bibframe/authority/BibframeAuthorityTypeProcessor.java:16]: Sonar is showing a violation in this line -> https://sonarcloud.io/project/issues?id=org.folio%3Amod-search&pullRequest=605&resolved=false&sinceLeakPeriod=true

**Review by PBobylev — APPROVED** [review]

**Review by pkjacob — APPROVED** [review]

File: src/main/resources/model/bibframe_authority.json [hunk 13]
```diff
@@ -0,0 +1,44 @@
+{
+  "name": "bibframe-authority",
+  "eventBodyJavaClass": "org.folio.search.domain.dto.BibframeAuthority",
+  "fields": {
+    "id": {
+      "index": "keyword"
+    },
+    "label": {
+      "index": "multilang"
```
Comment (psmagin) [inline src/main/resources/model/bibframe_authority.json:9]: "multilang" index will not properly as there is no language-defining field in the model.
You could use "standard" to get full-text capabilities.

File: src/main/resources/model/bibframe_authority.json [hunk 14]
```diff
@@ -0,0 +1,44 @@
+{
+  "name": "bibframe-authority",
+  "eventBodyJavaClass": "org.folio.search.domain.dto.BibframeAuthority",
```
Comment (psmagin) [inline src/main/resources/model/bibframe_authority.json:3]:
```suggestion
  "eventBodyJavaClass": "org.folio.search.domain.dto.BibframeAuthority",
  "reindexSupported": true
```
Should be defined to not break snapshot builds.

**Review by psmagin — CHANGES_REQUESTED** [review]

File: src/main/resources/model/bibframe_authority.json (hunk 10 above)
Comment (psmagin) [reply src/main/resources/model/bibframe_authority.json:31]: You still have same field "label" in both searchFields and Fields. If there is no special processing needed for the field, then I suggest to keep only one that is in Fields.

**Review by psmagin — APPROVED** [review]
